src/simsopt/mhd/spec.py

```diff
@@ -303,28 +974,74 @@ def run(self):
             logger.debug('About to call ending')
             spec.ending()
 
-        except:
+        except BaseException:
             if self.verbose:
                 traceback.print_exc()
             raise ObjectiveFailure("SPEC did not run successfully.")
 
         logger.info("SPEC run complete.")
-        # Barrier so workers do not try to read the .h5 file before it is finished:
+        # Barrier so workers do not try to read the .h5 file before it
+        # is finished:
         self.mpi.comm_groups.Barrier()
 
+        # Try to read SPEC output. 
         try:
             self.results = py_spec.SPECout(filename + '.sp.h5')
-        except:
+        except BaseException:
             if self.verbose:
                 traceback.print_exc()
-            raise ObjectiveFailure("Unable to read results following SPEC execution")
+            raise ObjectiveFailure(
+                "Unable to read results following SPEC execution")
 
         logger.info("Successfully loaded SPEC results.")
         self.need_to_run_code = False
 
+        # Deal with unconverged equilibria - these are excluded by 
+        # the optimizer, and the objective function is set to a large number
+        if self.results.output.ForceErr > self.tolerance:
+            raise ObjectiveFailure(
+                'SPEC could not find force balance'
+            )
+
+        # Save geometry as initial guess for next iterations
+        if update_guess:
+            try:
```

Comment:
It's best to keep `try` blocks as small as possible, especially when all exceptions are caught instead of just a specific exception, as is the case here. A lot happens in lines 1009-1028, so if an exception occurs, we don't know exactly what caused it. Can you move code outside of the `try` block so it encloses only the specific line that you expect to throw an exception?